pkgs/build-support/docker/stream_layered_image.py

```diff
@@ -130,6 +132,51 @@ def extract(self):
 LayerInfo = namedtuple("LayerInfo", ["size", "checksum", "path", "paths"])
 
 
+def add_base_layers(tar, from_image, mtime):
+    """
+    Adds the layers from the given base image to the final image.
+
+    tar: 'tarfile.TarFile' object for new layers to be added to.
+    from_image: Path to the base image archive.
+    mtime: 'mtime' of the added layer tarballs.
+    """
+    if from_image is None:
+        print("No 'fromImage' provided", file=sys.stderr)
+        return []
+
+    with tempfile.TemporaryDirectory() as base_layers:
+        tarfile.open(from_image).extractall(base_layers)
```

Comment:
Is there a way to do this without dumping the contents of `from_image` to the disk? I don't this `streamLayeredImage` should use the disk at all.

I think we can first read `manifest.json` directly from the tarball without extracing, then we can stream the layers from the tarball directly to the output. This should make extracting the input to a temporary directory unnecessary.

Reply:
The tarfile module presents a random-access interface to tarfiles, whereas tar files in general do not support random access, because they don't have an index and the data typically comes from a stream without random access (because of compression (or types I guess ;) ) ).
So if I understand the code correctly, we'll need two passes if we don't want to write to the filesystem. In the first pass, we only keep the `manifest.json` (probably easiest to use `TarFile.getmember`); in the second pass we stream the layers we need using TarFile.next() exclusively.
I'm basing this solely on the `tarfile` docs and general knowledge about tar and constant-memory streaming. Sadly the python docs don't document anything about performance or memory side effects, so if a lot of caching happens implicitly, `TarFile.getmember` may be a stupid recommendation. idk. `TarFile.next` seems like a low level method at least. I guess.

TBH I don't think the current read + write + read is much worse than read + read and optimizing it to be read + read will make the code a bit harder to follow.